runtime/parachains/src/runtime_api_impl/mod.rs

```diff
@@ -17,9 +17,13 @@
 //! Runtime API implementations for Parachains.
 //!
 //! These are exposed as different modules using different sets of primitives.
-//! At the moment there is only a v2 module and it is not completely clear how migration
-//! to a v2 would be done.
-
+//! At the moment there is a v2 module for the current stable api and
+//! vstaging module for all staging methods.
+//! When new version of the stable api is released it will be based on v2 and
+//! will contain methods from vstaging.
+//! The promotion consists of the following steps:
+//! 1. Bump the version of the stable module (e.g. v2 becomes v3)
+//! 2. Move methods from vstaging to v3. The new stable version can include some
+//!    or all methods from vstaging.
```

Comment:
Generally speaking I agree. Now after I've thought more about this I've got a concern.

I'm afraid we might have get to the point where we will want to release a new stable api version including methods from multiple staging methods. E.g.
```rust
#[api_version(2)]
trait MyApi {
    pub fn methodA();
    #[api_version(3)]
    pub fn methodB();
    #[api_version(4)]
    pub fn methodC();
}
```
Let say we want to release new stable api including both `methodB` and `methodC`, because we won't do such releases too frequently and we have stabilized both methods.

What we should do? Make v4 stable and skip v3? Or accept a convention where all staging apis will start from v100 (for example) so that we can freely decide what to include in the new stable versions?

My preference is for the latter.

Reply:
Well if we did intend to release them all at once, why didn't they all have the same version (3)? 

One thing that struck me when going through your PR though was that it might be useful to use constants, like:

`DISPUTES_API_VERSION` and use that both in `api_version()` calls and the using code. This way the version can change (e.g. we had to introduce an emergency version in between) without breaking any code.

Reply:
> Well if we did intend to release them all at once, why didn't they all have the same version (3)?

For example I work on `methodB`, you work on `methodC` a month after. They can both be released together but we put them in different version because I have already occupied v3 and you want to test runtime upgrade. Is this a valid case?

Or we can wipe the test network so that it is on its stable version (e.g. v2) and `methodC` goes in v3?

> One thing that struck me when going through your PR though was that it might be useful to use constants

I was thinking about the same. The version should be in a single place to avoid copy-paste errors like mine above.

Reply:
Good point, I might pick a different version for testing purposes, but once done I should set it back to 3 ... except if I really want to have it released later than your changes.

Reply:
> One thing that struck me when going through your PR though was that it might be useful to use constants

@eskimor unfortunately it's not possible to use a constant in `#[api_version()]` because it's a proc macro and at the point when the proc macro is expanded the compiler knows nothing about constants.

The second best I can do is to have the constant only on two places - in `api_version` and one more place (node/subsystem-types/src/messages.rs). This is how it looks like: https://github.com/paritytech/polkadot/pull/5567/commits/d52cc7460783cf89188b17e13bfea41be5d472b9

What do you think?

Reply:
I think the initial idea was to set the api_version to 100_000 + X or something like this for staging APIs. Once they are ready to be stabilized, we'd change their version to the next stable version. Versi and Rococo would always implement 100_000 + X version of the API, whereas polkadot and kusama would be on stable. 

Would downgrading a version (promoting to stable) create any breaking changes to testnets? 

Would could also have `Staging<Method>` in runtime api subsystem to distinguish staging APIs and rename them to `<Method>` ones they are stabilized.

Reply:
The nice thing about what we have right now, is that node side code does not need to change when a staging API becomes stable. It just queries the version and adjusts behaviour. If we pick some 100_000 version number, we would need to adjust code once it gets stable - especially since we can't use constants.

Reply:
> The second best I can do is to have the constant only on two places - in `api_version` and one more place (node/subsystem-types/src/messages.rs). This is how it looks like: [d2af967](https://github.com/paritytech/polkadot/commit/d2af967d4a00af4e2a2d185ee0da82e949bb2de1)
> 
> What do you think?

Looks like an improvement!